jax/scipy/linalg.py

```diff
@@ -264,3 +263,120 @@ def tril(m, k=0):
 @_wraps(scipy.linalg.triu)
 def triu(m, k=0):
   return np.triu(m, k)
+
+@jit
+def expm(A):
+    """
+    Compute the matrix exponential using Pade approximation.
+    Based on TensorFlow's tf.linalg.expm
+
+    Args:
+    A : square matrix of type float32, float64, complex64, complex128
+    
+    Returns:
+    Matrix exponential of A , square matrix of same dimensions as A
+        
+
+    See: "The Scaling and Squaring Method for the Matrix
+    Exponential Revisited", Nicholas Higham, 2005.
+    
+    TODO: Write a version to exploit the sparsity of A
+    """
+    A = np.asarray(A)
+    if A.ndim != 2 or A.shape[0] != A.shape[1]:
+        raise ValueError('expected A to be a square matrix')
+    A_L1 = np_linalg.norm(A,1)
+    n_squarings = 0
+    def _nest_where(vals, cases):
+      assert len(vals) == len(cases) - 1
+      if len(vals) == 1:
+        return np.where(
+            np.less(A_L1, vals[0]), cases[0], cases[1])
+      else:
+        return np.where(
+            np.less(A_L1, vals[0]), cases[0],
+            _nest_where(vals[1:], cases[1:]))
+    if A.dtype == 'float64' or A.dtype == 'complex128':
+       U3,V3 = _pade3(A)
+       U5,V5 = _pade5(A)
+       U7,V7 = _pade7(A)
+       U9,V9 = _pade9(A)
+       maxnorm = 5.371920351148152
+       n_squarings = np.maximum(0, np.floor_divide(np.log2(A_L1 / maxnorm),1))
+       A = A / 2**n_squarings
+       U13,V13 = _pade13(A)
+       conds=np.array([1.495585217958292e-002, 2.539398330063230e-001, 9.504178996162932e-001, 2.097847961257068e+000])
+       U=_nest_where(conds,(U3,U5,U7,U9,U13))
+       V=_nest_where(conds,(V3,V5,V7,V9,V13))
+    elif A.dtype == 'float32' or A.dtype == 'complex64':
+        U3,V3 = _pade3(A)
+        U5,V5 = _pade5(A)
+        maxnorm = 3.925724783138660
+        n_squarings = np.maximum(0, np.floor_divide(np.log2(A_L1 / maxnorm),1))
+        A = A / 2**n_squarings
+        U7,V7 = _pade7(A)
+        conds=np.array([4.258730016922831e-001, 1.880152677804762e+000])
+        U=_nest_where(conds,(U3,U5,U7))
+        V=_nest_where(conds,(V3,V5,V7))
+    else:
+        raise TypeError("A.dtype={} is not supported.".format(A.dtype))
+    P = U + V  # p_m(A) : numerator
+    Q = -U + V # q_m(A) : denominator
+    R = np_linalg.solve(Q,P)
```

Comment:
does the matrix `Q` have any special structure (e.g,. is it positive-definite)?

I guess probably not, since TensorFlow doesn't use it

Reply:
[Higham2005](http://eprints.ma.man.ac.uk/634/1/high05e.pdf): does not list any special structure for that `Q`.

Reply:
It looks like scipy includes some auto-detection for the case where A is upper triangular. We might eventually add this feature into JAX (or more realistically, we could add an option for indicating it explicitly).

Reply:
Yes, this is reasonable. I have managed to add in the explicit option.